test/plugins/test_plugins_handlers.py

```diff
@@ -100,3 +100,21 @@ def test_get_label_entrypoint(plugin_tmpfile):
     assert meta['label'] == 'plugin label'
     assert meta['type'] == handlers.EntryPointPlugin.PLUGIN_TYPE
     assert meta['source'] == 'test_plugin = file_mod'
+
+
+def test_entrypoint_plugin_get_version(plugin_tmpfile):
```

Comment:
This test case doesn't have any ` assert`. What is it checking?

Reply:
I'm not sure that it checks anything meaningful. When I couldn't break this new test case, even by modifying the code being tested to call `importlib.metadata.version()` with a nonexistent distribution name or to remove the try/except structure (in case it was catching an error and just `pass`ing), I got suspicious.

In the context of this test, the EntryPoint's `dist` attribute is `None`, so because of the `and hasattr(self.entry_point.dist, "name")` clause of the `if` statement edited in `EntryPointPlugin.get_version()`, that whole block is skipped and the return value is just the value from `super().get_version()`.

My conclusion is that this test doesn't set up the EntryPoint sufficiently. We can't meaingfully test "get the version of a plugin whose entry point name doesn't match its parent distribution's name" if the entry point isn't part of a distribution. It needs to also create a mock distribution, which I started looking into but ran out of time for.

Reply:
The check is that exceptions other than `ValueError` do not propagate beyond the version checking code.

Running the same test on `master` (2e8de023) produces the `PackageNotFoundError` that this test covers, because `master` has a too-narrow exception handler.

```
test/plugins/test_plugins_handlers.py:160: in test_entrypoint_plugin_get_version
    plugin.get_version()
sopel/plugins/handlers.py:627: in get_version
    version = importlib.metadata.version(self.module.__package__)
/home/jgerity/.pyenv/versions/3.8.14/lib/python3.8/importlib/metadata.py:530: in version
    return distribution(distribution_name).version
/home/jgerity/.pyenv/versions/3.8.14/lib/python3.8/importlib/metadata.py:503: in distribution
    return Distribution.from_name(distribution_name)
/home/jgerity/.pyenv/versions/3.8.14/lib/python3.8/importlib/metadata.py:177: in from_name
    raise PackageNotFoundError(name)
E   importlib.metadata.PackageNotFoundError: FAKEFAKEFAKE
```

However, I didn't remember any of this immediately from looking at the code again, so the test can be rewritten to make this more clear.

Reply:
I think @Exirel's point is that the test only verifies the absence of an exception, rather than also verifying that correct version metadata is fetched for this case.

Ideally we'd have several test cases related to this:

1. The "normal" case: Distribution name and package/module name match (e.g. `sopel_testplugin` installable with `pip install sopel_testplugin`)
   Assert: plugin version matches the distribution version.
2. Distribution name and package name don't match (e.g. `sopel_myplugins.test` installable with `pip install my_sopel_plugins`), where `sopel_myplugins.test` has its own version that would be returned by `super().get_version()`
   Assert: plugin version matches the version of `sopel_myplugins.test`
3. Same as 2, but `sopel_myplugins.test` does _not_ have its own version string so we fall through to getting the distribution's version instead.
   Assert: plugin version matches the distribution version
4. Same as 3, but `entry_point` does not have a `dist` (or the `dist` does not have a `name`).
   Assert: plugin version is `None`; all attempts to get one from the module or distribution failed

As the test is currently written, the code that would have raised `PackageNotFoundError` on `master` is simply skipped over. The `if` statement now checks `hasattr(self.entry_point, "dist") and hasattr(self.entry_point.dist, "name")`, which evaluates as `False` because no distribution is mocked up; it's a bare entry point with no parent.

That's a fine case to check, but we also need at least some behavior tests where the distribution-related conditions actually pass, and a (mock) distribution's metadata can be inspected.

Reply:
> rather than also verifying that correct version metadata is fetched for this case.

See 7d6c7b1 which adds an assertion, if not a particularly strong one.

I agree that there are probably several tests for write for **the feature**, but this PR is concerned primarily with (2), the bug it is meant to fix. In other words, this is a regression test to cover the bug I found and *only* this case.

Reply:
I'm fine with the PR not testing everything. However, I still want a test that represent an existing use-case, and describe what output is expected in said use case (i.e. what would be the result of `get_version()`).

Looking at the latest diff, it looks like you added the appropriate assert, and so I am happy!